python_ta/transforms/type_inference_visitor.py

```diff
@@ -248,6 +248,14 @@ def _assign_type(self, target: NodeNG, expr_type: type) -> None:
                     for subtarget in target.elts:
                         target_tvar = self.lookup_type(subtarget, subtarget.name)
                         self.type_constraints.unify(target_tvar, contained_type)
+        elif isinstance(target, astroid.Subscript):
+            # TODO: previous case must recursively handle this one
+            set_call_type = self._handle_call(target, '__setitem__', target.value.inf_type.getValue(),
+                              target.slice.inf_type.getValue(), expr_type)
+            if isinstance(set_call_type, TypeFail):
+                target.inf_type = set_call_type
+            else:
+                target.inf_type = TypeInfo(expr_type)
```

Comment:
This behaviour isn't consistent with the other branches, in which `target.inf_type` isn't changed (from `NoType`). It's okay if you want to record the failure since we've made those changes above in `visit_assign`, but don't change the "success" case here.

Reply:
But if we have an assignment like `a[3] = 0` (say `a` is not homogeneous), where would we store the fact that the expression `a[3]` has type `int` if not in the subscript node (which being the target of the assignment would still have `NoType` if not explicitly unified with `expr_type`)?
Perhaps I should have a lookup function like in the other cases of `_assign_type` and then unify with the corresponding type variable?

Reply:
@simeonkr I've merged that in, so please go ahead. You'll need to fix a merge conflict, which hopefully will be pretty minor.